fiducial_follow/follower_controller.py

```diff
@@ -0,0 +1,153 @@
+from custom_messages.msg import FollowerCommand, FollowerStatus
+
+from math import pi, sqrt, atan2
+import traceback
+import math
+import time
+
+
+class Controller:
+    """
+    Constructor for our class
+    """
+    def __init__(self):
+       rospy.init_node('controller')
+
+       # Time per loop for the main control
+       self.loop_msec = 50
+
+       # A publisher for sending commands to the follower node
+       self.followerCmdPub = rospy.Publisher("/follower_commands", FollowerCommand, queue_size=1)
+
+       # Subscribe to incoming status 
+       rospy.Subscriber("/follower_status", FollowerStatus, self.newFollowerStatus)
+
+       # define a fiducial that when set implies we are not seeking any fiducial
+       self.null_fiducial = "fid0"
+
+       # ---------------------- start of keywords for use in commands ---------------------------
+
+       # Defines that are of value to pass in as elements of messages in the inbound command topic
+
+       # default action we will use once we actually find and have approached the fiducial
+       # The awaitNextCommand action will continue to follow until a new command comes in
+       self.actOnDoneStopFollowing = "StopFollowing"
+       self.actOnDoneDoNextCommand = "DoNextCommand"
+
+       # We need a state to indicate for the current command so we define 'InProgress' and 'Done'
+       # When cmdStatus is InProgress we do not read new command except state control commands
+       self.cmdStatusInProgress = "InProgress"
+       self.cmdStatusDone       = "Done"
+
+       # define names for specific commands 
+       self.cmdFollowFiducial  = "FollowFiducial"
+       self.cmdClearCommands   = "ClearCommands"
+       self.cmdClearInProgress = "ClearInProgress"
+       self.cmdStopMovement    = "StopMovement"
+       self.cmdDriveForward    = "DriveForward"
+       self.cmdDriveReverse    = "DriveReverse"
+       self.cmdRotateLeft      = "RotateLeft"
+       self.cmdRotateRight     = "RotateRight"
+
+    def publishFollowerCommand(self, cmdType, actionOnDone, string1, numParam1, comment):
+        cmdMsg = FollowerCommand()
+        cmdMsg.commandType   = cmdType
+        cmdMsg.actionOnDone  = actionOnDone
+        cmdMsg.strParam1     = string1
+        cmdMsg.numParam1     = numParam1
+        cmdMsg.comment       = comment
+        self.followerCmdPub.publish(cmdMsg)
+
+    """
+    Called when a Tracker Command is received
+    """
+    def newFollowerStatus(self, msg):
+        commandType  = msg.commandType
+        statusState  = msg.statusState
+        string1      = msg.string1
+        param1       = msg.num1
+        string2      = msg.string2
+        param2       = msg.num2
+        print "New Follower status of Type: %s with statusState '%s' and str1 '%s' \n" % \
+            (commandType, statusState, string1)
+
+
+    """
+    Main loop
+    """
+    def run(self):
+        # setup for looping at 20hz
+        rate = rospy.Rate(self.loop_msec)
```

Comment:
`rospy.Rate()` takes a frequency in Hz, so if `loop_msec` is in milliseconds, the argument should be `1000.0 / loop_msec`

Reply:
I tried to make the frequency of the loop exactly like before but only have it be in a variable.

Reply:
The units of that variable would be Hz not me.